catalog/sql/sql.go

```diff
@@ -0,0 +1,721 @@
+	"github.com/uptrace/bun/schema"
+)
+
+type SupportedDialect string
+
+const (
+	Postgres SupportedDialect = "postgres"
+	MySQL    SupportedDialect = "mysql"
+	SQLite   SupportedDialect = "sqlite"
+	MSSQL    SupportedDialect = "mssql"
+	Oracle   SupportedDialect = "oracle"
+)
+
+const (
+	DialectKey           = "sql.dialect"
+	DriverKey            = "sql.driver"
+	initCatalogTablesKey = "init_catalog_tables"
+)
+
+func init() {
+	catalog.Register("sql", catalog.RegistrarFunc(func(name string, p iceberg.Properties) (c catalog.Catalog, err error) {
+		driver, ok := p[DriverKey]
+		if !ok {
+			return nil, errors.New("must provide driver to pass to sql.Open")
+		}
+
+		dialect := strings.ToLower(p[DialectKey])
+		if dialect == "" {
+			return nil, errors.New("must provide sql dialect to use")
+		}
+
+		uri := strings.TrimPrefix(p.Get("uri", ""), "sql://")
+		sqldb, err := sql.Open(driver, uri)
+		if err != nil {
+			return nil, err
+		}
+
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("failed to create SQL catalog: %v", r)
+			}
+		}()
+
+		return NewCatalog(p.Get(name, "sql"), sqldb, SupportedDialect(dialect), p)
+	}))
+}
+
+var (
+	_ catalog.Catalog = (*Catalog)(nil)
+)
+
+var (
+	minimalNamespaceProps = iceberg.Properties{"exists": "true"}
+
+	dialects  = map[SupportedDialect]schema.Dialect{}
+	dialectMx sync.Mutex
+)
+
+func createDialect(d SupportedDialect) schema.Dialect {
+	switch d {
+	case Postgres:
+		return pgdialect.New()
+	case MySQL:
+		return mysqldialect.New()
+	case SQLite:
+		return sqlitedialect.New()
+	case MSSQL:
+		return mssqldialect.New()
+	case Oracle:
+		return oracledialect.New()
+	default:
+		panic("unsupported sql dialect")
+	}
+}
+
+func getDialect(d SupportedDialect) schema.Dialect {
+	dialectMx.Lock()
+	defer dialectMx.Unlock()
+	ret, ok := dialects[d]
+	if !ok {
+		ret = createDialect(d)
+		dialects[d] = ret
+	}
+	return ret
+}
+
+type sqlIcebergTable struct {
+	bun.BaseModel `bun:"table:iceberg_tables"`
+
+	CatalogName              string `bun:",pk"`
+	TableNamespace           string `bun:",pk"`
+	TableName                string `bun:",pk"`
+	MetadataLocation         sql.NullString
+	PreviousMetadataLocation sql.NullString
+}
+
+type sqlIcebergNamespaceProps struct {
+	bun.BaseModel `bun:"table:iceberg_namespace_properties"`
+
+	CatalogName   string `bun:",pk"`
+	Namespace     string `bun:",pk"`
+	PropertyKey   string `bun:",pk"`
+	PropertyValue sql.NullString
+}
+
+func withReadTx[R any](ctx context.Context, db *bun.DB, fn func(context.Context, bun.Tx) (R, error)) (result R, err error) {
+	db.RunInTx(ctx, &sql.TxOptions{ReadOnly: true}, func(ctx context.Context, tx bun.Tx) error {
+		result, err = fn(ctx, tx)
+		return err
+	})
+	return
+}
+
+func withWriteTx(ctx context.Context, db *bun.DB, fn func(context.Context, bun.Tx) error) error {
+	return db.RunInTx(ctx, &sql.TxOptions{Isolation: sql.LevelLinearizable}, func(ctx context.Context, tx bun.Tx) error {
```

Comment:
double checking, is `LevelLinearizable` what we want here?

Reply:
It's what made the most sense to me, unless you think I should just use `LevelDefault` and let the underlying DB decide?